servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/P2CSelector.java

```diff
@@ -0,0 +1,136 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.loadbalancer;
+
+import io.servicetalk.client.api.LoadBalancedConnection;
+import io.servicetalk.client.api.LoadBalancer;
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.context.api.ContextMap;
+
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.Predicate;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.concurrent.api.Single.succeeded;
+
+/**
+ * This {@link LoadBalancer} selection algorithm is based on work by Michael David Mitzenmacher in The Power of Two
+ * Choices in Randomized Load Balancing.
+ *
+ * @see <a href="https://www.eecs.harvard.edu/~michaelm/postscripts/tpds2001.pdf">Mitzenmacher (2001) The Power of Two
+ * Choices in Randomized Load Balancing</a>
+ */
+final class P2CSelector<ResolvedAddress, C extends LoadBalancedConnection>
+        extends BaseHostSelector<ResolvedAddress, C> {
+
+    @Nullable
+    private final Random random;
+    private final int maxEffort;
+
+    P2CSelector(final String targetResource, final int maxEffort, @Nullable final Random random) {
+        super(targetResource);
+        this.maxEffort = maxEffort;
+        this.random = random;
+    }
+
+    @Override
+    public Single<C> selectConnection(
+            @Nonnull List<Host<ResolvedAddress, C>> hosts,
+            @Nonnull Predicate<C> selector,
+            @Nullable ContextMap context,
+            boolean forceNewConnectionAndReserve) {
+        final int size = hosts.size();
+        switch (size) {
+            case 0:
+                // We shouldn't get called if the load balancer doesn't have any hosts.
+                throw new AssertionError("Selector for " + getTargetResource() +
+                        " received an empty host set");
+            case 1:
+                Host<ResolvedAddress, C> host = hosts.get(0);
+                if (!forceNewConnectionAndReserve) {
+                    C connection = host.pickConnection(selector, context);
+                    if (connection != null) {
+                        return succeeded(connection);
+                    }
+                }
+                // Either we require a new connection or there wasn't one already established so
+                // try to make a new one if the host is healthy. If it's not healthy, we fail
+                // and let the higher level retries decide what to do.
+                if (!host.isActiveAndHealthy()) {
+                    return noActiveHostsException(hosts);
```

Comment:
nit: this isn't an exception it's a failed `Single`. I might expect `return noActiveHosts(hosts)`.

Reply:
Sorry, I didn't see this until I had merged. I'll clean it up in a follow up.